File: Dynamic-Programming/322-M-Coin-Change/coin_change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        OPT = [float('inf')] * (amount + 1)
        OPT[0] = 0

        for i in range(1, amount + 1):
            for coin in coins:
                if coin <= i:
                    OPT[i] = min(OPT[i], 1 + OPT[i - coin])

        if OPT[amount] == float('inf'):
            return -1
            
        return OPT[amount]

File: Dynamic-Programming/322-M-Coin-Change/coin_change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int n = coins.size();
        vector<int> OPT(amount + 1, amount+1);
        OPT[0] = 0;
        
        for (int i=1; i<amount+1; i++) {
            for (const auto& coin : coins) {
                if (coin <= i)
                    OPT[i] = min(OPT[i], OPT[i-coin] + 1);
            }
        }
        if (OPT[amount] == amount+1) {
            return -1;
        }
        return OPT[amount];
    }
};